src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableClient.cs

```diff
@@ -491,12 +491,29 @@ async Task<PurgeHistoryResult> IDurableOrchestrationClient.PurgeInstanceHistoryA
             return this.DurabilityProvider.GetOrchestrationStateWithPagination(condition, cancellationToken);
         }
 
+        private static EntityQueryResult ConvertToEntityQueryResult(IEnumerable<DurableEntityStatus> entities, string continuationToken)
+        {
+            var entityQueryResult = new EntityQueryResult();
+            entityQueryResult.Entities = entities;
+            entityQueryResult.ContinuationToken = continuationToken;
+
+            return entityQueryResult;
+        }
+
         /// <inheritdoc />
         async Task<EntityQueryResult> IDurableEntityClient.ListEntitiesAsync(EntityQuery query, CancellationToken cancellationToken)
         {
             var condition = new OrchestrationStatusQueryCondition(query);
-            var result = await ((IDurableClient)this).ListInstancesAsync(condition, cancellationToken);
-            var entityResult = new EntityQueryResult(result);
+            EntityQueryResult entityResult;
+
+            do // make sure that the returned page contains at least one element, if there is at least one element
+            {
+                var result = await ((IDurableClient)this).ListInstancesAsync(condition, cancellationToken);
+                entityResult = new EntityQueryResult(result, query.IncludeDeleted);
+                condition.ContinuationToken = entityResult.ContinuationToken;
+            }
+            while (entityResult.ContinuationToken != null && !entityResult.Entities.Any());
```

Comment:
I'm trying to think about the case when `entityResult.ContinuationToken != null` is true and `!entityResult.Entities.Any()` is false. I think that means there are more records to read but we're going to break early anyways because we found at least one entity. Is that the intended design? If so, why wouldn't we keep reading the remaining records to see if we can find more entities? It seems to me that users will never get more than one page of results back.

Reply:
Random 2¢. I think as an application developer, I wouldn't want it churning through pages of results to return me a single full page of results. That could really hurt performance outside of my control. If it only does one page (even if it's only one result), then I can timebox it (return as many results as possible in 100ms) or whatever to return results applicable to my application.

Reply:
Thanks for chiming in, @withinboredom. Now that you mention this, I realize that I misunderstood what's happening here. I forgot that both `EntityQueryResult` and `EntityQuery` have continuation token fields, meaning that it's perfectly fine to return a single page of results because the caller can continue to ask for more pages if they want. This should have been obvious by looking more closely at the previous implementation, but with this in mind, I think this solution is perfectly reasonable (pages may contain just one entity in the worst case, but you can get more).

Timeboxing - i.e. limiting the amount of time we spend churning through pages for the first entity, is another interesting feature, but I think that could be done as a separate improvement in a new PR (it might require interface changes, though).

Reply:
Those are good points. 

As written right now, the code could take arbitrary time to return the first result (for example, if lots of entities were deleted and AzureStorage keeps returning those first). In such a situation I should return an empty result and a continuation token to keep the client request from timing out.